node/src/components/chain_synchronizer/operations.rs

```diff
@@ -351,125 +357,184 @@ impl CanUseSyncingNodes for Deploy {}
+            T::TAG,
+            id,
+            peer
+        );
+        match ctx.effect_builder.fetch::<T>(id, peer).await {
+            Ok(fetched_data @ FetchedData::FromStorage { .. }) => {
+                trace!(
+                    "did not get {:?} with id {:?} from {:?}, got from storage instead",
+                    T::TAG,
+                    id,
+                    peer
+                );
+                return Some(Ok(fetched_data));
+            }
+            Ok(fetched_data @ FetchedData::FromPeer { .. }) => {
+                trace!("fetched {:?} with id {:?} from {:?}", T::TAG, id, peer);
+                return Some(Ok(fetched_data));
+            }
+            Err(FetcherError::Absent { .. }) => {
+                warn!(
+                    ?id,
+                    tag = ?T::TAG,
+                    ?peer,
+                    "chain sync could not fetch; trying next peer",
+                );
+                ctx.mark_bad_peer(peer);
+            }
+            Err(FetcherError::TimedOut { .. }) => {
+                warn!(
+                    ?id,
+                    tag = ?T::TAG,
+                    ?peer,
+                    "peer timed out",
+                );
+                ctx.mark_bad_peer(peer);
+            }
+            Err(error @ FetcherError::CouldNotConstructGetRequest { .. }) => {
+                return Some(Err(error.into()))
+            }
+        }
+    }
+    None
+}
+
 enum TrieAlreadyPresentOrDownloaded {
     AlreadyPresent,
     Downloaded(Bytes),
 }
 
-async fn fetch_trie_retry_forever<REv>(
+async fn fetch_trie_with_retries<REv>(
     id: Digest,
     ctx: &ChainSyncContext<'_, REv>,
 ) -> Result<TrieAlreadyPresentOrDownloaded, FetchTrieError>
 where
     REv: From<FetcherRequest<TrieOrChunk>> + From<NetworkInfoRequest>,
 {
-    let trie_or_chunk =
-        match fetch_retry_forever::<_, TrieOrChunk>(ctx, TrieOrChunkId(0, id)).await? {
-            FetchedData::FromStorage { .. } => {
-                return Ok(TrieAlreadyPresentOrDownloaded::AlreadyPresent)
-            }
-            FetchedData::FromPeer {
-                item: trie_or_chunk,
-                ..
-            } => *trie_or_chunk,
-        };
+    let result = fetch_with_retries::<_, TrieOrChunk>(ctx, TrieOrChunkId(0, id)).await;
+    let trie_or_chunk = match result {
+        Ok(FetchedData::FromStorage { .. }) => {
+            return Ok(TrieAlreadyPresentOrDownloaded::AlreadyPresent)
+        }
+        Ok(FetchedData::FromPeer {
+            item: trie_or_chunk,
+            ..
+        }) => *trie_or_chunk,
+        Err(FetchWithRetryError::AttemptsExhausted {
+            id,
+            total_attempts,
+            attempts_after_bootstrapped,
+        }) => {
+            error!(
+                total_attempts,
+                attempts_after_bootstrapped,
+                ?id,
+                "fetch attempts exhausted"
+            );
+            return Err(FetchTrieError::AttemptsExhausted);
```

Comment:
If we log this inside `fetch_from_peers` (assuming it's not logged on the call side) then we can remove this match and propagate the error immediately with
```rust
fetch_with_retries::<_, TrieOrChunk>(ctx, TrieOrChunkId(0, id)).await?;
```

Reply:
Error handling was reworked in 20e06ed73831fa8b42d7bd782f125b52848772a5.
In particular, we now log the "fetch attempts exhausted" only in `fetch_with_retries()`.

It's not perfect, but I tend to leave it as is, again because the error structure will eventually be reworked in https://github.com/casper-network/casper-node/issues/3287.

Reply:
Thanks. That looks better :)